### ISSUE TITLE
Unable to get Windows to read certain deflated files

### ISSUE THREAD

**Issue by pcrockett** [mvdnes/zip-rs#99]
If I create a zip file with a subdirectory, Windows won't properly read that subdirectory. Any loose files sitting in the root of the archive can be extracted just fine, but trying to extract a file in a subdirectory yields an error.

The following will generate a zip file that reproduces this behavior:

```rust
use std::fs::{ OpenOptions };
use std::io::Write;
use zip::{ CompressionMethod, ZipWriter };
use zip::write::FileOptions;

fn main() {
    let archive = OpenOptions::new()
        .create_new(true)
        .write(true)
        .open("temp.zip")
        .unwrap();

    let mut writer = ZipWriter::new(archive);

    let options = FileOptions::default()
        .compression_method(CompressionMethod::Deflated);
    writer.start_file("noproblem.txt", options).unwrap();
    writer.write_all(b"Windows can extract this file just fine.").unwrap();

    let options = FileOptions::default();
    writer.add_directory("subdir", options).unwrap();

    let options = FileOptions::default()
        .compression_method(CompressionMethod::Deflated);
    writer.start_file("subdir/problem.txt", options).unwrap();
    writer.write_all(b"Windows will show an error when trying to extract this file.").unwrap();

    writer.flush().unwrap();
}
```

And here is the error I'm encountering when trying to drag / drop the `subdir/problem.txt` file onto my Desktop:

![Unexpected Error](https://user-images.githubusercontent.com/1114385/53703719-906feb80-3e15-11e9-96f4-bea942134bc4.png)

Running Windows 10 Home edition. 7zip extracts this just fine.

I recognize this may be a problem with Windows' built-in zip file support, but then again I've never encountered a zip file that _didn't_ work out of the box with Windows.

**Comment by mvdnes** [comment]
This is a very strange issue indeed! I have not been able to find the exact cause just yet, but it does not seem to be a related to it being a file in a subdirectory. If you change the contents of the subdir/noproblem.txt file to

```rust
writer.write_all(b"Windows will show an error when trying to extract this file.\r\nHowever, it has no problems with it after this modification.\r\n").unwrap();
```

then it will extract just fine. The issue also seems to be gone when changing the compression method to `Stored`.
It seems that it might be a problem with the compression. Some further investigation is definitely needed.

**Comment by mvdnes** [comment]
So I have been investigating this issue, and the problem seems to be in the deflate stream of certain small files. I will try to get it resolved with the people of libflate.

Technical info:
It seems that Windows does not like a flate stream with zero distance codes. When manually adding a distance code table without modifying the LZ77 stream the file is perfectly readable on Windows. The RFC does seem to allow an empty distance table, so I am not sure who is at fault.

**Comment by mvdnes** [comment]
This should be fixed in version 0.5.1.
Thanks for the report, and if you still have issues feel free to reopen this.

**Comment by pcrockett** [comment]
Awesome. Thanks for getting that fixed, and so fast! Fist bump! 🤜 🤛 

_(and thanks for the library itself)_
